flatgfa/src/packedseq.rs

```diff
@@ -0,0 +1,328 @@
+        match value {
+            Nucleotide::A => 0,
+            Nucleotide::C => 1,
+            Nucleotide::T => 2,
+            Nucleotide::G => 3,
+        }
+    }
+}
+
+impl From<Nucleotide> for char {
+    fn from(value: Nucleotide) -> Self {
+        match value {
+            Nucleotide::A => 'A',
+            Nucleotide::C => 'C',
+            Nucleotide::G => 'G',
+            Nucleotide::T => 'T',
+        }
+    }
+}
+
+/// A compressed vector-like structure for storing nucleotide sequences
+///     - Two base pairs are stored per byte
+///
+/// data: a vector that stores a compressed encoding of this vec's sequence
+/// high_nibble_end: true if the final base pair in the sequence is stored at a
+///                   high nibble
+struct PackedVec {
+    data: Vec<u8>,
+    high_nibble_end: bool,
+}
+
+impl PackedVec {
+    fn new() -> Self {
+        PackedVec {
+            data: Vec::new(),
+            high_nibble_end: true,
+        }
+    }
+    pub fn push(&mut self, input: Nucleotide) {
+        let value = input.into();
+        assert!(value <= 0xF);
+        if self.high_nibble_end {
+            self.data.push(value);
+            self.high_nibble_end = false;
+        } else {
+            let last_index = self.data.len() - 1;
+            self.data[last_index] = (value << 4) | self.data[last_index];
+            self.high_nibble_end = true;
+        }
+    }
+}
+
+impl fmt::Display for PackedVec {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[")?;
+        let iter = PackedVecIterator::new(&self);
+        let mut i = 0;
+        for item in iter {
+            if i == 0 {
+                i = 1;
+            } else {
+                write!(f, ", ");
+            }
+            println!("u8!!! {}", item);
+            let n: Nucleotide = item.into();
+            let c: char = n.into();
+            println!("CHAR!!! {}", c);
+            write!(f, "{}", c)?;
+        }
+        write!(f, "]")
+    }
+}
+
+struct PackedVecIterator<'a> {
+    data: &'a Vec<u8>,
+    cur_index: usize,
+    cur_high_nibble: bool,
+    high_nibble_end: bool,
+    last: bool,
+}
+
+impl<'a> PackedVecIterator<'a> {
+    pub fn new(vec: &'a PackedVec) -> Self {
+        Self {
+            data: &vec.data,
+            cur_index: 0,
+            cur_high_nibble: false,
+            high_nibble_end: vec.high_nibble_end,
+            last: false,
+        }
+    }
+}
+
+impl<'a> Iterator for PackedVecIterator<'a> {
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cur_index >= self.data.len() || self.last {
+            return None;
+        }
+        if self.cur_index == self.data.len() - 1 && !self.high_nibble_end {
+            self.last = true;
+        }
+        let result = if self.cur_high_nibble {
+            (self.data[self.cur_index] >> 4) & 0b00001111u8
+        } else {
+            self.data[self.cur_index] & 0b00001111u8
+        };
```

Comment:
Per the above comment, this duplicates the logic in `get_vec_elem`. Using a simpler iterator approach, you can reuse a single implementation of this bit-slicing code.